command/agent/consul/client.go

```diff
@@ -0,0 +1,678 @@
+	exec driver.ScriptExecutor, task *structs.Task) error {
+
+	id := makeTaskServiceID(allocID, task.Name, service)
+	host, port := task.FindHostAndPortFor(service.PortLabel)
+	serviceReg := &api.AgentServiceRegistration{
+		ID:      id,
+		Name:    service.Name,
+		Tags:    make([]string, len(service.Tags)),
+		Address: host,
+		Port:    port,
+	}
+	// copy isn't strictly necessary but can avoid bugs especially
+	// with tests that may reuse Tasks
+	copy(serviceReg.Tags, service.Tags)
+	ops.regServices = append(ops.regServices, serviceReg)
+
+	for _, check := range service.Checks {
+		checkID := createCheckID(id, check)
+		if check.Type == structs.ServiceCheckScript {
+			if exec == nil {
+				return fmt.Errorf("driver doesn't support script checks")
+			}
+			ops.scripts = append(ops.scripts, newScriptCheck(
+				allocID, task.Name, checkID, check, exec, c.client, c.logger, c.shutdownCh))
+
+		}
+		host, port := serviceReg.Address, serviceReg.Port
+		if check.PortLabel != "" {
+			host, port = task.FindHostAndPortFor(check.PortLabel)
+		}
+		checkReg, err := createCheckReg(id, checkID, check, host, port)
+		if err != nil {
+			return fmt.Errorf("failed to add check %q: %v", check.Name, err)
+		}
+		ops.regChecks = append(ops.regChecks, checkReg)
+	}
+	return nil
+}
+
+// RegisterTask with Consul. Adds all sevice entries and checks to Consul. If
+// exec is nil and a script check exists an error is returned.
+//
+// Actual communication with Consul is done asynchrously (see Run).
+func (c *ServiceClient) RegisterTask(allocID string, task *structs.Task, exec driver.ScriptExecutor) error {
+	ops := &operations{}
+	for _, service := range task.Services {
+		if err := c.serviceRegs(ops, allocID, service, exec, task); err != nil {
+			return err
+		}
+	}
+	c.commit(ops)
+	return nil
+}
+
+// UpdateTask in Consul. Does not alter the service if only checks have
+// changed.
+func (c *ServiceClient) UpdateTask(allocID string, existing, newTask *structs.Task, exec driver.ScriptExecutor) error {
+	ops := &operations{}
+
+	existingIDs := make(map[string]*structs.Service, len(existing.Services))
+	for _, s := range existing.Services {
+		existingIDs[makeTaskServiceID(allocID, existing.Name, s)] = s
+	}
+	newIDs := make(map[string]*structs.Service, len(newTask.Services))
+	for _, s := range newTask.Services {
+		newIDs[makeTaskServiceID(allocID, newTask.Name, s)] = s
+	}
+
+	parseAddr := newTask.FindHostAndPortFor
+
+	// Loop over existing Service IDs to see if they have been removed or
+	// updated.
+	for existingID, existingSvc := range existingIDs {
+		newSvc, ok := newIDs[existingID]
+		if !ok {
+			// Existing sevice entry removed
+			ops.deregServices = append(ops.deregServices, existingID)
+			for _, check := range existingSvc.Checks {
+				ops.deregChecks = append(ops.deregChecks, createCheckID(existingID, check))
+			}
+			continue
+		}
+
+		// Service exists and wasn't updated, don't add it later
+		delete(newIDs, existingID)
```

Comment:
Totally missed the fact that port changes are excluded from service and check IDs, so great catch. Added a huge test to exercise port changes and did lots of fixing: https://github.com/hashicorp/nomad/pull/2467/commits/970979d8fa31119b425fda35e866b00861f55108